git-cliff-core/src/gitlab.rs

```diff
@@ -0,0 +1,449 @@
+
+/// Log message to show when done fetching from GitLab.
+pub const FINISHED_FETCHING_MSG: &str = "Done fetching GitLab data.";
+
+/// Returns the GitLab API url either from environment or from default value.
+fn get_gitlab_api_url() -> String {
+	env::var(GITLAB_API_URL_ENV)
+		.ok()
+		.unwrap_or_else(|| GITLAB_API_URL.to_string())
+}
+
+/// Trait for handling the different entries returned from the GitLab API.
+trait GitLabEntry {
+	/// Returns the API URL for fetching the entries at the specified page.
+	fn url(id: i64, repo: &str, owner: &str, page: i32) -> String;
+	/// Returns the request buffer size.
+	fn buffer_size() -> usize;
+}
+
+/// https://docs.gitlab.com/ee/api/projects.html#get-single-project
+/// Representation of a single GitLab Project.
+#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct GitLabProject {
+	/// GitLab id for project
+	pub id:                  i64,
+	/// Optional Description of project
+	pub description:         Option<String>,
+	/// Name of project
+	pub name:                String,
+	/// Name of project with namespace owner / repo
+	pub name_with_namespace: String,
+	/// Name of project with namespace owner/repo
+	pub path_with_namespace: String,
+	/// Project created at
+	pub created_at:          String,
+	/// Default branch eg (main/master)
+	pub default_branch:      String,
+}
+
+impl GitLabEntry for GitLabProject {
+	fn url(_id: i64, repo: &str, owner: &str, _page: i32) -> String {
+		format!("{}/projects/{}%2F{}", get_gitlab_api_url(), owner, repo)
+	}
+	fn buffer_size() -> usize {
+		1
+	}
+}
+
+/// https://docs.gitlab.com/ee/api/commits.html
+/// Representation of a single commit.
+#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct GitLabCommit {
+	/// Sha
+	pub id:              String,
+	/// Short Sha
+	pub short_id:        String,
+	/// Git message
+	pub title:           String,
+	/// Author
+	pub author_name:     String,
+	/// Author Email
+	pub author_email:    String,
+	/// Authored Date
+	pub authored_date:   String,
+	/// Committer Name
+	pub committer_name:  String,
+	/// Committer Email
+	pub committer_email: String,
+	/// Committed Date
+	pub committed_date:  String,
+	/// Created At
+	pub created_at:      String,
+	/// Git Message
+	pub message:         String,
+	/// Parent Ids
+	pub parent_ids:      Vec<String>,
+	/// Web Url
+	pub web_url:         String,
+}
+
+impl GitLabEntry for GitLabCommit {
+	fn url(id: i64, _repo: &str, _owner: &str, page: i32) -> String {
+		let commit_page = page + 1;
+		format!(
+			"{}/projects/{}/repository/commits?per_page={MAX_PAGE_SIZE}&\
+			 page={commit_page}",
+			get_gitlab_api_url(),
+			id
+		)
+	}
+	fn buffer_size() -> usize {
+		10
+	}
+}
+
+/// https://docs.gitlab.com/ee/api/merge_requests.html
+/// Representation of a single pull request.
+#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct GitLabMergeRequest {
```

Comment:
Is it fine to keep all the fields from the GitLab response? What happens if the API changes, or some of the fields are `null`? Then it might sense to make those `Option<T>`. 

What do you think? Is this an issue?

Reply:
Can we remove the fields that are not strictly needed? You can take a look at GitHub model for getting an idea!

Reply:
@orhun done. I have cut down the fields.